Add support of unnamed labels with @ (.localchar) prefix

vrubleg opened the pull request:
All local labels in ca65 use `@` prefix (or another symbol that can be selected using `.LOCALCHAR`). Unnamed symbols don't follow this rule and use inconsistent, hard to highlight, and hard to notice syntax with a mere `:` as a label (that is obviously not consistent with the `.FEATURE labels_without_colons +` setting) and weird looking `:+/:-` as a reference.

I propose to fix all these issues by introduction a new optional syntax where unnamed labels are defined by `@:`, and referenced by `@+/@-` or `@<` or `@>`. This new syntax is consistent with other local labels and follow the `.LOCALCHAR` and `.FEATURE labels_without_colons +` settings. Also it is easy to highlight in text editors, unnamed labels naturally get color of local labels.

Ticket #2128

bbbradsmith commented:
My biggest concern with this PR is there is no test case which would both demonstrate how the feature is intended to work, and also verify that it does.

I understand the desire to want `@:` as an alias for `:`, and think this is an OK addition by itself.

However I don't think adding new aliases `<` and `>` for `-` and `+` have the same utility, and also they should not be special case for `@:` only. Either we should have them for both `:` and `@` or neither. (My vote is neither.)

Another question I have is whether `@-` searches the preceding local scope for _any_ label beginning with `@`, or whether it only includes `@:` or does it also include `:`?

I think intuitively I would expect `@-` to include all `@` labels, anonymous or not, and exclude the anonymous `:` that doesn't start with `@`, and correspondingly `:-` should include only `:` as it does currently, and not attach to `@:` as well. What does this PR do for these cases?

mrdudz commented:
> However I don't think adding new aliases < and > for - and + have the same utility, and also they should not be special case for @: only. Either we should have them for both : and @ or neither. (My vote is neither.)

I agree, that should be removed from this PR

vrubleg commented:
In the current PR, the `@:` and `:` are 100% synonyms, they just create an anonymous label.

The `:+`/`:-` and `@+`/`@-` or `@>`/`@<` are also complete synonyms and refer to a next or a previous anonymous label, so all non-anonymous `@labels` are not considered.

It is easy to add support of `<` and `>` into `:`, it is also easy to remove them from `@`.

vrubleg commented:
Making it a setting will complicate code somewhat, but if it is the only way to go, I can do this. In this case, I could propose two separate settings:

1. The style of an anonymous label definition (`@:` or `:`) that will also affect the first character of a reference (`@` or `:`).
```
.FEATURE localchar_anonymous_labels +
```
2. The style of reference counter (`</>` or `-/+`).
```
.FEATURE angle_bracket_anonymous_refs +
```

vrubleg commented:
Updated PR:
- Fixed a small issue in default VS project settings that I made previous time. `WholeProgramOptimization` was in wrong place and didn't work. Also added a few flags to generate more clean executables without [Rich headers](https://www.ntcore.com/files/richsign.htm).
- Added basic `.editorconfig` so Visual Studio follows some of conventions that are used in this project automatically.
- Fixed msbuild.bat to correctly find Visual Studio when it's not in default path.
- Added `.FEATURE line_continuations +` that fixes #2133 (moves `.LINECONT` into `.FEATURE` where all the other settings live, for consistency).
- The unnamed labels can be referenced also by `:<` and `:>`, in addition to all previous variants.

All of these features are in separate commits for convenience.

vrubleg commented:
Updated the docs to reflect the changes. I didn't make the feature settings to configure what style of unnamed labels should be used. All of them are accepted at the same time (with equal meaning) for better compatibility.

vrubleg commented:
Separated this MR into three PRs. #2204 is already merged, #2209 is pending. After it is merged, I'll rebase the current #2132 to include just changes related to the unnamed labels improvements.

mrdudz commented:
> Separated this PR into three PRs. #2204 is already merged, #2209 is pending. After it is merged, I'll rebase the current #2132 to include just changes related to the unnamed labels improvements.

Was that done? Just wondering if i should see this in the log here or not :)

vrubleg commented:
Rebased this branch. Now it includes just the unnamed label related changes =)

mrdudz commented:
This one really needs a test... The test should make sure the different variants actually work, and also that a mixture of the variants work (@+ and @> should be really synonymous).

Other than that i am willing to merge it :)

vrubleg commented:
The test was pushed a month ago =)

mrdudz commented:
Sorry - please always post something when a PR is ready/modified - i don't get notified when something is pushed to the branch only... merging
